### PR TITLE
feat: add context in EmbeddingFunction

### PR DESCRIPTION
**Pull request by bjwswang**
Fix: #8

### CODE REVIEW

**Comment by bjwswang** [comment]
@tazarov  Do you mind if I change the `EmbeddingFunction`  https://github.com/amikos-tech/chroma-go/blob/main/chroma.go#L19 to  adapt langchain `Embedder`

Then we can keep consistent with langchaingo (Langchaingo seems to be the 1st grade customer of chroma-go now and in the near future)

In langchaingo, a converter func  https://github.com/tmc/langchaingo/blob/main/vectorstores/chroma/embedder.go#L13 will be no longer required.

**Comment by bjwswang** [comment]
@tazarov btw... the langchaingo author asked if you have any release plan on chroma-go.

**Comment by tazarov** [comment]
@bjwswang, what do you mean by release plan? Do they need a specific issue (e.g., this one) merged/fixed?

**Comment by bjwswang** [comment]
> @bjwswang, what do you mean by release plan? Do they need a specific issue (e.g., this one) merged/fixed?

I think the author is willing to utilize a stable version of chroma-go.  They do not have any specific issue to be merged right now .

File: cohere/cohere.go [hunk 1]
```diff
@@ -130,13 +135,12 @@ func (e *CohereEmbeddingFunction) CreateEmbedding(documents []string) ([][]float
 	return response.Embeddings, nil
 }
 
-func (e *CohereEmbeddingFunction) CreateEmbeddingWithModel(documents []string, model string) ([][]float32, error) {
+func (e *CohereEmbeddingFunction) EmbedQuery(ctx context.Context, document string) ([]float32, error) {
 	response, err := e.apiClient.CreateEmbedding(&CreateEmbeddingRequest{
-		Model: model,
-		Texts: documents,
+		Texts: []string{document},
```
Comment (tazarov) [inline cohere/cohere.go:140]: let's also pass the context to CreateEmbedding and then pass it to the HTTP request

File: hf/hf.go [hunk 2]
```diff
@@ -122,13 +127,12 @@ func (e *HuggingFaceEmbeddingFunction) CreateEmbedding(documents []string) ([][]
 	return response.Embeddings, nil
 }
 
-func (e *HuggingFaceEmbeddingFunction) CreateEmbeddingWithModel(documents []string, model string) ([][]float32, error) {
-	e.apiClient.Model = model
+func (e *HuggingFaceEmbeddingFunction) EmbedQuery(ctx context.Context, document string) ([]float32, error) {
 	response, err := e.apiClient.CreateEmbedding(&CreateEmbeddingRequest{
```
Comment (tazarov) [inline hf/hf.go:131]: let's also pass the context to CreateEmbedding and then pass it to the HTTP request

File: openai/client.go [hunk 3]
```diff
@@ -160,7 +165,7 @@ func ConvertToMatrix(response *CreateEmbeddingResponse) [][]float32 {
 	return matrix
 }
 
-func (e *OpenAIEmbeddingFunction) CreateEmbedding(documents []string) ([][]float32, error) {
+func (e *OpenAIEmbeddingFunction) EmbedDocuments(ctx context.Context, documents []string) ([][]float32, error) {
 	response, err := e.apiClient.CreateEmbedding(&CreateEmbeddingRequest{
```
Comment (tazarov) [inline openai/client.go:169]: let's also pass the context to CreateEmbedding and then pass it to the HTTP request

File: openai/client.go [hunk 4]
```diff
@@ -174,16 +179,16 @@ func (e *OpenAIEmbeddingFunction) CreateEmbedding(documents []string) ([][]float
 	return ConvertToMatrix(response), nil
 }
 
-func (e *OpenAIEmbeddingFunction) CreateEmbeddingWithModel(documents []string, model string) ([][]float32, error) {
+func (e *OpenAIEmbeddingFunction) EmbedQuery(ctx context.Context, document string) ([]float32, error) {
```
Comment (tazarov) [inline openai/client.go:182]: let's also pass the context to CreateEmbedding and then pass it to the HTTP request

**Review by tazarov — CHANGES_REQUESTED** [review]
@bjwswang, can you take a look at a couple of improvements and one nit I suggested?

File: hf/hf.go [hunk 5]
```diff
@@ -100,6 +103,8 @@ func (c *HuggingFaceClient) CreateEmbedding(req *CreateEmbeddingRequest) (*Creat
 	return &createEmbeddingResponse, nil
 }
 
+var _ chroma.EmbeddingFunction = (*HuggingFaceEmbeddingFunction)(nil)
```
[the remark this answers is not in the document]
Comment (bjwswang) [reply hf/hf.go:106]: This is to make sure `HuggingFaceEmbeddingFunction ` really implements interface `chroma.EmbeddingFunction `

File: cohere/cohere.go (hunk 1 above)
Comment (bjwswang) [reply cohere/cohere.go:140]: Done

File: hf/hf.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tazarov) [reply hf/hf.go:106]: Pardon my ignorance, but isn't implementing all interface methods what makes an implementation?

I also removed the line, and this doesn't affect the outcome.

**Comment by tazarov** [comment]
@bjwswang, can you rebase from main? I did a fix on the workflow so that tests will succeed. Also take a look at my last comment.

File: hf/hf.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bjwswang) [reply hf/hf.go:106]: @tazarov Not at all.  You have done a great job! 

I have asked gpt in poe . please see [this link](https://poe.com/s/uPTBTTtnUF07ulcFnLoI)

**Comment by bjwswang** [comment]
> @bjwswang, can you rebase from main? I did a fix on the workflow so that tests will succeed. Also take a look at my last comment.

Done

**Pull request merged**